client: add dedicated inode lock support for libcephfs

lxbsz opened the pull request:
This patch will add one dedicated mutex lock for each Inode
 object. For some members in Inode class, such as the ino, snapid
 and faked_ino, there is no need to be protected by the inode_lock
 because they are readonly once the inode object is initilized.
    
 The Inode::inode_locks could be recursive, but the order must be:
    
 ```
parent->inode_lock.lock()
 {
   child->inode_lock.lock()
   {
     grandson->inode_lock.lock()
     {
       ...
     }
     grandson->inode_lock.unlock()
   }
   child->inode_lock.unlock()
 }
 parent->inode_lock.unlock()
```
    
 For the Dentry class, it will also be protected by the Inode it
 belong to. And in the Inode::inode_lock scope it allows to hold
 the client_lock when needed, but not vise versa.
  
 To avoid different Inode object get the same lock id in lockdep,
 it will use the vino to build a unique lock name.
 
 There has one exception for Inode::iget() and Inode::iput(), they
 must put under the client_lock scope. That also means for the
 InodeRef type variable must be assigned under the client_lock scope
 too. This will avoid one race between releasing the Inode memory
 and getting the nref.
    
1, for the XXXRef related patches, they are mainly to hold the instances to make sure they won't be released during the client_lock's unlock/lock gap.

2, for some members in the Inode/Fh, etc, there is no need to protect them by the inode_lock or client_lock, because they won't change once initilized in the whole life time, this could help simplify the inode_lock patch.

3, the inode_lock will protect almost all the members in the Inode class, except the nref inherited from the RefCountedObject and it will be protected by the client_lock, this could make sure that during the client_lock and inode_lock's unlock/lock gap, the nref will be consistent to fix #40028.

4, the minor cleaning patches for the Inode/Fh, etc classes, just to simplify the code to make it easier to read and do not need to care too much about the members order when initilizing them.

5, for the osdc related changes, mainly added the oc_lock support to protect the members in the ObjectCacher class, and move the outside locks, like the inode_lock, to the ObjectSet class.

6, I have tried to break the big patches into small ones as possible as I can, but for the last two ones, they are still very big.

7, and also please note that the lockdep will affect the performance because it will try to construct/destrcut the Backtrace very frequently, the more inode locks the poorer performance will it be. This won't be a problem, in the build release the lockdep will always be disabled. Just use the lockdep for developing or test cases to help debug the bugs.



 Fixes: https://tracker.ceph.com/issues/46688
 Signed-off-by: Xiubo Li <xiubli@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

lxbsz commented:
This PR depends on https://github.com/ceph/ceph/pull/39774 and https://github.com/ceph/ceph/pull/39742.

lxbsz commented:
jenkins retest this please

lxbsz commented:
jenkins test api

lxbsz commented:
jenkins test api

dillaman commented on src/librbd/cache/ObjectCacherObjectDispatch.cc:
```diff
@@ -128,7 +128,7 @@ void ObjectCacherObjectDispatch<I>::init() {
                 << " max_dirty_age=" << max_dirty_age << dendl;
 
   m_object_cacher = new ObjectCacher(cct, m_image_ctx->perfcounter->get_name(),
-                                     *m_writeback_handler, m_cache_lock,
```
Nit: `ObjectCacheObjectDispatch` and `ObjectCacherWriteback` have assumptions about `m_cache_lock`. Now while I am fine with removing it, these assumptions in `librbd` need to be fixed. Preferably by removing `ImageCtx::m_cache_lock` and adding a new lock to `ObjectCacherWriteback` for protecting the data structures.

lxbsz replied:
Sure, will check it.

lxbsz replied:
@dillaman 
Done, have made the m_lock a private one in `ObjectCacherWriteback`.

lxbsz replied:
And the `./bin/ceph_test_librbd` test has succeeded. Is that enough ?

lxbsz commented:
@batrick This is ready to review. Thanks.

lxbsz commented:
jenkins test make check

lxbsz commented:
Without this PR:
```
[root@lxbceph1 cephfs]# time git clone /data/ceph/.git
Cloning into 'ceph'...
done.
Updating files: 100% (10475/10475), done.

real    5m21.802s
user    0m2.712s
sys     0m16.009s
```

With this PR:
```
[root@lxbceph1 cephfs]# time git clone /data/ceph/.git
Cloning into 'ceph'...
done.
Updating files: 100% (10475/10475), done.

real    4m30.621s
user    0m2.755s
sys     0m18.144s

```
Test serval times, the performance improved around 15%.

But the lockdep must be disabled as we do in the release build, because there have many inode locks and then the lockdep backtrace code will eat too many memories and may cause OOM, and at the same time the lockdep will affect the perf too, anyway we need to improve the lockdep code.

lxbsz commented:
> But the lockdep must be disabled as we do in the release build, because there have many inode locks and then the lockdep backtrace code will eat too many memories and may cause OOM, and at the same time the lockdep will affect the perf too, anyway we need to improve the lockdep code.

@batrick

For the `bin/ceph_test_libcephfs` and most teuthology tests, the lockdep will work well, but for the test cases who are trying to create too many files, like clone the `ceph` repo and do some tests, it may cause OOM. Because the Backtraces for lockdep will eat too many memories.

Can we just switch to use the std::mutex instead of the ceph::mutex ? The lockdep will also degrade the performance sharply. I have tried to cache the Backtrace objects to prevent them being allocated/freed frequently, it helps but not much as needed.

Any idea ?

jtlayton reviewed (commented):
Fleshing out the commit logs would be nice here. What does doing this give you? It may seem obvious now, but later folks will be thankful.

lxbsz commented:
> Fleshing out the commit logs would be nice here. What does doing this give you? It may seem obvious now, but later folks will be thankful.

Done. And add some brief describtion about the patch set here two:

1, for the XXXRef related patches, the are mainly to hold the instances to make sure they won't be released during the `client_lock`'s unlock/lock gap.

2, for some members in the Inode/Fh, etc, there is no need to protect them by the `inode_lock` or `client_lock`, because they won't change once initilized in the whole life time, this could help simplify the `inode_lock` patch.

3, the `inode_lock` will protect almost all the members in the Inode class, except the `nref` inherited from the `RefCountedObject` and it will be protected by the `client_lock`,  this could make sure that during the `client_lock` and `inode_lock`'s unlock/lock gap, the `nref` will be consistent to fix #40028.

4, the minor cleaning patches for the Inode/Fh, etc classes, just to simplify the code to make it easier to read and do not need to care too much about the members order when initilizing them.

5, for the osdc related changes, mainly added the `oc_lock` support to protect the members in the `ObjectCacher` class, and move the outside locks, like the `inode_lock`, to the `ObjectSet` class.

6, I have tried to break the big patches into small ones as possible as I can, but for the last two ones, they are still very big.

Thanks.

lxbsz commented:
jenkins test make check

lxbsz commented:
The make check failure is not related:

```
INFO  2021-03-17 04:32:28,124 [shard 0] ms - test_v2_protocol: start local TestPeer at v2:127.0.0.1:9013/0...
ERROR 2021-03-17 04:32:28,125 [shard 0] ms - FailoverTestPeer: there is another instance running at v2:127.0.0.1:9013/0
ERROR 2021-03-17 04:32:28,127 [shard 0] none - ../src/test/crimson/test_messenger.cc:1623 : In function '{anonymous}::FailoverTestPeer::init(entity_addr_t)::<lambda(const std::error_code&)>', abort(%s)
abort() called
Aborting on shard 0.
Backtrace:
  /usr/lib/x86_64-linux-gnu/libasan.so.5+0x000000000006bb2f
  0x000000000d78b358
  0x000000000d76369b
```

Just rebased to the lastest. Will rebase again after that is fixed.

jtlayton commented on src/client/Client.cc:
```diff
@@ -7203,10 +7204,11 @@ int Client::mkdir(const char *relpath, mode_t mode, const UserPerm& perm, std::s
     return r;
+
+  std::scoped_lock lock(client_lock);
   if (cct->_conf->client_permissions) {
```
Commenting here because github doesn't allow comments against the changelog. The patch makes sense, but the commit log is bogus. This is not a "cleanup" per-se, but rather a lock pushdown, where we're reducing the footprint of the client_lock by having a lower-level function take it itself.

Aside from that, we should note that you're increasing the mutex thrashing here. Previously, we'd take the client lock once and just release it when Client::mkdir returns. Now you take it in path_walk, release it when it returns, and take it again here for the rest of mkdir. That's generally less efficient than it was before. If this is an interim change that gets cleaned up later then that's ok. Otherwise, we'll need to weigh whether this change is worthwhile.

lxbsz replied:
This is an interim change, and preparing for the inode lock. Without this we need to release the `client_lock` in the `path_walk` to take the  `inode_lock` and take the client_lock again, like we do in the kclient:

```
client_lock.unlock();
std::scoped_lock il{in->inode_lock};
client_lock.lock();
```

lxbsz replied:
This is an interim change, and preparing for the inode lock. Without this we need to release the `client_lock` in the `path_walk` to take the  `inode_lock` and take the client_lock again, like we do in the kclient:

```
client_lock.unlock();
std::scoped_lock il{in->inode_lock};
client_lock.lock();
```

lxbsz replied:
I add some logs and have found that ususally it will take 80ns in average to return from acquiring the mutex locks.

Without the inode lock patches, the `client_lock` ususally will wait 15ms in average if it needs to wait, with this PR it will reduce to 8ms in average.

So the thrashing should be acceptable ?

This is not that precise, I just add some logs randomly in the code to get the results.

lxbsz replied:
Cool, this is what I can do to split the patch into small ones, or all something like this must be squashed into the inode lock one.

jtlayton commented on src/client/Client.cc:
```diff
@@ -10097,8 +10097,10 @@ int Client::pwritev(int fd, const struct iovec *iov, int iovcnt, int64_t offset)
+				   bool clamp_to_int)
 {
+    std::unique_lock cl(client_lock);
+
```
Nice cleanup. I have to wonder why this ever took a mutex argument in the first place.

lxbsz replied:
The new PR: https://github.com/ceph/ceph/pull/40204

jtlayton commented on src/include/lru.h:
```diff
@@ -189,6 +189,10 @@ class LRU {
+  bool is_on_lru(LRUObject *o) {
+    return o->lru_link.is_on_list();
+  }
+
```
Typically it's best to introduce new functions like this along with their first caller, so that we can see how it's intended to be used. Maybe squash this patch together with the first one that uses is_on_lru()?

lxbsz replied:
> Typically it's best to introduce new functions like this along with their first caller, so that we can see how it's intended to be used.

For this, the caller is in the inode lock patch, and only when the inode lock is introduced will the caller is needed. This is just to make sure during the `client_lock`'s unlock/lock gap, we won't release or change the object twice.

> Maybe squash this patch together with the first one that uses is_on_lru()?

Sure.

jtlayton commented on src/client/Client.h:
```diff
@@ -1065,6 +1065,7 @@ class Client : public Dispatcher, public md_config_obs_t {
    */
   void _finish_init();
 
+  ceph::spinlock ll_snap_ref_lock;
```
Userland spinlocks are somewhat unpredictable as you can get scheduled off the CPU while holding one, in which case you can end up taking a rather nasty performance hit -- other locking threads will still be spinning on the CPUs until the holder is scheduled again and can release it. In userland, the general wisdom is to use mutexes instead of spinlocks unless you have a *very* good reason. I'm not sure this qualifies.

lxbsz replied:
> Userland spinlocks are somewhat unpredictable as you can get scheduled off the CPU while holding one, in which case you can end up taking a rather nasty performance hit -- other locking threads will still be spinning on the CPUs until the holder is scheduled again and can release it.

Yeah, sounds reasonable.

> In userland, the general wisdom is to use mutexes instead of spinlocks unless you have a _very_ good reason. I'm not sure this qualifies.

While the `ll_snap_ref_lock` spin lock here will only protect `ll_snap_ref` map and the critical section is very simple and short. The case you mentioned should be okay ?

lxbsz replied:
@jtlayton 

I am also thinking whether will it make sense when the `fromdir` and `todir` have no parent/child relationship, can we just use the `Inode::vino` to order the locks, let's always make the smaller one in front the other one when taking the lock ? 

The `Inode::vino` won't change in Inode's life time, then we can get rid of the `s_vfs_rename_mutex` global lock.

Could you foresee any potential issue for this ?

jtlayton replied:
Simple and small is good for a spinlock-protected critical, but again, it can be hard to predict how they will perform. I generally recommend avoiding them in userland code unless you *really* know what you're doing.

Basing the locking on the Inode::vino does sound reasonable. I'll think about it a bit more though.

lxbsz replied:
> Simple and small is good for a spinlock-protected critical, but again, it can be hard to predict how they will perform. I generally recommend avoiding them in userland code unless you _really_ know what you're doing.
> 

Yeah, you are right. 
I think it is hard to predict what the `map/list/vector, etc` internally will do, maybe allocating new memories which could cause sleeping.

> Basing the locking on the Inode::vino does sound reasonable. I'll think about it a bit more though.

Sure, thanks.
Locally the rename lock patch is ready and can reduce many lock thrashing :-)

jtlayton commented on src/client/Client.cc:
```diff
@@ -7180,11 +8003,15 @@ int Client::rename(const char *relfrom, const char *relto, const UserPerm& perm,
+        return r;
+    }
+
+    std::scoped_lock il(todir->inode_lock);
```
Wait, so you don't need to hold the locks for both fromdir and todir during a cross-directory rename? Why not?

If it turns out that you do need to hold both locks here, then you have an issue. You really can't rely on parent/child relationships to determine the ordering since the fromdir and todir may not be nested.

FWIW, the kernel does this by looking at the value of the dentry pointer and the lower one goes first. You may need to do something similar here instead of trying to rely on parent/child ordering for the locking.

lxbsz replied:
Yeah, this need to be improved, I am still thinking how to make it work well if `todir` and `fromdir` are not be nested.

Currently there no case will the inode locks will nest unless they have the parent/child relation.

Checked the kernel code, it seems it must hold the global `s_vfs_rename_mutex` before there has inode locks nesting to avoid the dead lock. This sounds a good idea.

jtlayton replied:
Yeah, it's not terribly scalable (i.e., it probably ought to be a per-sb mutex), but that's what we have in the kernel for now.

lxbsz replied:
Yeah, locally the `rename lock` patch is ready and can reduce many lock thrashing :-)

lxbsz replied:
The logic is similar:
1, if `fromdir` equals to `todir`, it's simple, just take the `fromdir`'s inode lock

For case1, there is no need to take the global `s_vfs_rename_mutex`.

2, if `fromdir` is ancester, then the lock order will be `fromdir->inode_lock` --> `todir->inode_lock`
3, if `todir` is ancester, then the lock order will be `todir->inode_lock` --> `fromdir->inode_lock`
4, or take the `inode_locks` in any order will be okay.

For case2~4, it will always take the global `s_vfs_rename_mutex` first.

Here the case2 and case3 are necessary and will always keep the lock order if they have parent/child relationship. Then it could concurrent run with other threads where the inode locks may nest, and won't cause the dead lock.

jtlayton reviewed (changes requested):
Ok, went through this in some more detail this morning and added some more specific comments. In general though, I think this PR needs to be broken up into multiple smaller PRs. Probably:

1) lockdep changes
2) cleanup/initialization patches (limit this one to things that shouldn't change behavior at all and can be easily reviewed)
3) osdc changes (if it's possible to separate those out)
4) the client_lock breakup and inode_lock work

lxbsz commented:
> Ok, went through this in some more detail this morning and added some more specific comments. In general though, I think this PR needs to be broken up into multiple smaller PRs. Probably:
> 
Okay, I will try to split it into some small PRs as possible.

> 1. lockdep changes
This has already split out and merged.

> 2. cleanup/initialization patches (limit this one to things that shouldn't change behavior at all and can be easily reviewed)
This is okay.

> 3. osdc changes (if it's possible to separate those out)
> 4. the client_lock breakup and inode_lock work
The osdc changes couldn't be seprated from the inode lock patch, at least the build will fail. Actually the last two patches in the above patch set should be folded. I just split it to make it easier to review.

lxbsz commented on src/client/FhRef.h:
```diff
@@ -0,0 +1,12 @@
+class Fh;
+void intrusive_ptr_add_ref(Fh *fh);
+void intrusive_ptr_release(Fh *fh);
+typedef boost::intrusive_ptr<Fh> FhRef;
```
> Hmm...just today, @batrick gave a talk on new pointer types, shared_ptr in particular. Should we be expanding the use of intrusive pointers here? Does this give us something that the new std:: pointers don't?
> 

I will check it to see which ones could switch to the shared ptr, but for the InodeRef/CapRef couldn't.

> It would be nice too if there were more explanation of the changes in the changelogs. 

I will add more detailed explaination about it.

>This patch series overall just jumps into making a lot of these changes without explaining why and what the benefit of these changes are. I can imagine that we do need a FhRef object, but it would be nice to understand why it's necessary and how this relates to the BCL (Big Client Lock).

Okay, I will add it.

lxbsz commented:
Have split two https://github.com/ceph/ceph/pull/40183 and https://github.com/ceph/ceph/pull/39742, will try to split more tomorrow.
